### PR TITLE
kube-apiserver: split apart generic control plane options

### PR DESCRIPTION
**Pull request by sttts**
/kind feature

This PR splits up the kube-apiserver options into
1. the generic part in `pkg/controlplane/apiserver/options`
2. the kube specific part in `cmd/kube-apiserver/app/options`.

Functionally, nothing changes. This PR only splits up and moves code, in preparation of [KEP-4080 generic contral plane staging repo](https://github.com/kubernetes/enhancements/pull/4052), i.e. the package location is not final.

```release-note
NONE
```

### CODE REVIEW

File: cmd/kube-apiserver/app/options/completion.go [hunk 1]
```diff
@@ -19,141 +19,78 @@ package options
...
 func Complete(opts *ServerRunOptions) (CompletedOptions, error) {
-	// set defaults
-	if err := opts.GenericServerRunOptions.DefaultAdvertiseAddress(opts.SecureServing.SecureServingOptions); err != nil {
-		return CompletedOptions{}, err
+	if opts == nil {
+		return CompletedOptions{completedOptions: &completedOptions{}}, nil
 	}
 
-	// process s.ServiceClusterIPRange from list to Primary and Secondary
+	// process options.ServiceClusterIPRange from list to Primary and Secondary
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/completion.go:53]: nit: opts

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -23,45 +23,30 @@ import (
...
-	logsapi "k8s.io/component-base/logs/api/v1"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/cluster/ports"
+	controlplaneapiserver "k8s.io/kubernetes/pkg/controlplane/apiserver/options"
 	"k8s.io/kubernetes/pkg/controlplane/reconcilers"
-	_ "k8s.io/kubernetes/pkg/features" // add the kubernetes feature gates
+	_ "k8s.io/kubernetes/pkg/features"
 	kubeoptions "k8s.io/kubernetes/pkg/kubeapiserver/options"
+
+	// add the kubernetes feature gates
```
Comment (sttts) [inline cmd/kube-apiserver/app/options/options.go:34]: undo import changes

Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go:34]: done

File: pkg/controlplane/apiserver/.import-restrictions
Comment (dims) [inline pkg/controlplane/apiserver/.import-restrictions:1]: +1 to let the higher up directory's .import-restrictions kick in.

File: cmd/kube-apiserver/app/options/options.go [hunk 3]
```diff
@@ -23,45 +23,29 @@ import (
...
-	Metrics                 *metrics.Options
-	Logs                    *logs.Options
-	Traces                  *genericoptions.TracingOptions
+	*controlplaneapiserver.Options // embedded to avoid noise in existing consumers
+	CloudProvider                  *kubeoptions.CloudProviderOptions
 
+	Extra
+}
+
+type Extra struct {
```
Comment (dims) [inline cmd/kube-apiserver/app/options/options.go:45]: nit...`Extra` ... could we do something better for the name?

File: pkg/controlplane/apiserver/options/options_test.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (dims) [inline pkg/controlplane/apiserver/options/options_test.go:2]: Let's update dates?

File: cmd/kube-apiserver/app/options/options.go (hunk 3 above)
Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go:45]: we use to call these `Extra` everywhere else. `Values` would be an alternative, or we could embed it. But I would prefer to do that globally in a follow-up if we want to.

Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go:45]: Nvmd. We are embedding.

File: cmd/kube-apiserver/app/options/completion.go [hunk 5]
```diff
@@ -19,141 +19,78 @@ package options
...
+	kubeoptions "k8s.io/kubernetes/pkg/kubeapiserver/options"
 )
 
 // completedOptions is a private wrapper that enforces a call of Complete() before Run can be invoked.
 type completedOptions struct {
-	*ServerRunOptions
+	controlplane.CompletedOptions
+	CloudProvider *kubeoptions.CloudProviderOptions
+
+	Extra
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/kube-apiserver/app/options/completion.go:38]: as above.

File: pkg/controlplane/apiserver/options/options_test.go (hunk 4 above)
Comment (sttts) [reply pkg/controlplane/apiserver/options/options_test.go:2]: done

File: cmd/kube-apiserver/app/options/options.go (hunk 3 above)
Comment (dims) [reply cmd/kube-apiserver/app/options/options.go:45]: some discussion on slack https://kubernetes.slack.com/archives/C0EG7JC6T/p1687809103644619?thread_ts=1687785734.583899&cid=C0EG7JC6T

**Pull request merged**